File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Senior.java

package com.watabou.pixeldungeon.actors.mobs;

import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Stun;
import com.watabou.utils.Callback;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class Senior extends Monk {

	{
		spriteClass = "spritesDesc/Senior.json";
		dmgMin = 12;
		dmgMax = 20;
	}
	
	@Override
	public int attackProc(@NotNull Char enemy, int damage ) {
		if (Random.Int( 10 ) == 0) {
			Buff.prolong( enemy, Stun.class, 1.1f );
		}
		return super.attackProc( enemy, damage );
	}
	
	@Override
	public boolean actMeleeAttack(Char enemy) {
		if (Random.Float() < 0.3f) {
			// Play kick animation with callback to trigger attack logic
			getSprite().playExtra("kick", new Callback() {
				@Override
				public void call() {
					// After kick animation completes, perform the actual attack
					if (isAlive()) {
						onAttackComplete();
					}
				}
			});
			getSprite().turnTo(getPos(), enemy.getPos());
			spend(attackDelay());
			return false;
		} else {
			return super.actMeleeAttack(enemy);
		}
	}
}

File: scripts/mobs/Senior.lua
--[[
  Senior - monk elder, City boss floor trash (batch 17b, was
  actors/mobs/Senior.java). Monk stats with a heavier hit and a 1/10
  stun proc on top of the disarm. Kick anim on 3/10 attacks - same
  story as Monk: the java override sat on the hero-only
  actMeleeAttack path; here it rides doAttack and plays.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

local function kickAttack(self, enemy, chance)
    if enemy == nil or math.random() >= chance then
        return false
    end

    self:setEnemy(enemy)
    self:spend(self:attackDelay())
    self:getSprite():turnTo(self:getPos(), enemy:getPos())

    if RPD.Dungeon:isCellVisible(enemy:getPos()) and not RPD.Dungeon:realtime() then
        RPD.CharUtils:extraAttack(self, "kick")
    else
        self:onAttackComplete()
    end
    return true
end

return mob.init{
    doAttack = function(self, enemy)
        return kickAttack(self, enemy, 0.3)
    end,
    attackProc = function(self, enemy, dmg)
        if enemy ~= nil then
            if math.random(10) == 1 then
                RPD.Buffs.Buff:prolong(enemy, "Stun", 1.1)
            end
            RPD.CharUtils:disarm(self, enemy)
        end
        return dmg
    end
}
